examples/mining_cfd.py

```diff
@@ -0,0 +1,125 @@
+    return result
+
+
+def get_supported_mask(pattern, table):
+    table_height = table.shape[0]
+    def get_mask(cfd_item):
+        if cfd_item.value is None:
+            true_array = np.full(
+                shape=table_height,
+                fill_value=True,
+                dtype=np.bool_
+            )
+            return pandas.Series(true_array)
+        columnName = table.columns[cfd_item.attribute]
+        return table[columnName].astype(str).str.lower() == cfd_item.value.lower()
+
+    return reduce(operator.and_, map(get_mask, pattern))
+
+
+# returns a dictionary
+#   key: a tuple of attribute values (all attributes from lhs)
+#   value: indices of all rows with those attribute values from supported_df
+def make_lhs_to_row_nums(lhs, supported_df):
+    lhs_to_row_nums = defaultdict(list)
+    lhs_column_indices = items_to_indices(lhs)
+    for row_num, row_series in supported_df.iterrows():
+        key = tuple(row_series.iloc[lhs_column_indices])
+        lhs_to_row_nums[key].append(row_num)
+    return lhs_to_row_nums
+
+
+def validate_cfd(lhs, rhs, table):
+    supported_mask = get_supported_mask(lhs, table)
+    supported_df = table[supported_mask]
+    lhs_to_row_nums = make_lhs_to_row_nums(lhs, supported_df)
+    rows_satisfying_cfd = []
+
+    # add rows with identical lhs to rows_satisfying_cfd if they have the most frequent rhs
+    for lhs_tuple, row_nums_with_lhs in lhs_to_row_nums.items():
+        # fill rhs_to_row_nums
+        rhs_to_row_nums = defaultdict(list)
+        for row_num in row_nums_with_lhs:
+            key = table.iloc[row_num, rhs.attribute]
+            rhs_to_row_nums[key].append(row_num)
+
+        best_count = 0
+        row_nums_with_most_frequent_rhs = None
+        for rhs_key, row_nums_with_rhs in rhs_to_row_nums.items():
+            if len(row_nums_with_rhs) > best_count:
+                best_count = len(row_nums_with_rhs)
+                row_nums_with_most_frequent_rhs = row_nums_with_rhs
+        rows_satisfying_cfd.extend(row_nums_with_most_frequent_rhs)
+
+    return rows_satisfying_cfd, supported_df.index
+
+
+def visualize_cfd(cfd, table):
+    print('CFD:')
+    print(cfd, ":\n")
+
+    rows_satisfying_cfd, supported_rows = validate_cfd(cfd.lhs_items, cfd.rhs_item, table)
+
+    header, *rows = table.to_string().splitlines()
+    print(DEFAULT_BG_CODE, header)
+    for index, row in enumerate(rows):
+        if index not in supported_rows:
+            color_code = DEFAULT_BG_CODE
+        elif index in rows_satisfying_cfd:
+            color_code = GREEN_BG_CODE
+        else:
+            color_code = RED_BG_CODE
+        print(color_code, row, DEFAULT_BG_CODE)
+
+    support = len(supported_rows)
+    num_rows_satisfy_cfd = len(rows_satisfying_cfd)
+    print("lhs count: ", len(cfd.lhs_items))
+    print("support: ", support, "\033[1;42m \033[1;41m \033[1;49m")
+    print("confidence: \033[1;32m", num_rows_satisfy_cfd, "\033[1;37m/", support,
```

Comment:
nit: hard-coded values (some of them have constants already), something like `f'{GREEN_BG_CODE} {RED_BG_CODE} {DEFAULT_BG_CODE}'` is preferred (though I'm not sure if this is supposed to be here, doesn't this just change the background where there is no text?).

Reply:
purely visual thing to show that support is the amount of red and green lines, fixed to be global constants